shell/platform/fuchsia/dart_runner/dart_component_controller_v1.cc

```diff
@@ -0,0 +1,165 @@
+    return;
+  }
+  if (controller.is_valid()) {
+    binding_.Bind(std::move(controller));
+    binding_.set_error_handler([this](zx_status_t status) { Kill(); });
+  }
+}
+
+DartComponentControllerV1::~DartComponentControllerV1() {}
+
+void DartComponentControllerV1::Kill() {
+  Shutdown();
+}
+
+void DartComponentControllerV1::Detach() {
+  binding_.set_error_handler([](zx_status_t status) {});
+}
+
+void DartComponentControllerV1::SendReturnCode() {
+  binding_.events().OnTerminated(return_code_,
+                                 fuchsia::sys::TerminationReason::EXITED);
+}
+
+fdio_ns_t* DartComponentControllerV1::PrepareNamespace() {
+  fdio_ns_t* ns = nullptr;
+  zx_status_t status = fdio_ns_create(&ns);
+  if (status != ZX_OK) {
+    return nullptr;
+  }
+
+  fuchsia::sys::FlatNamespace* flat = &startup_info_.flat_namespace;
+
+  dart_utils::RunnerTemp::SetupComponent(ns);
+
+  for (size_t i = 0; i < flat->paths.size(); ++i) {
+    if (flat->paths.at(i) == kComponentTmpPath) {
+      // /tmp is covered by the local memfs.
+      continue;
+    }
+
+    zx::channel dir;
+    if (flat->paths.at(i) == kServiceRootPath) {
+      // clone /svc so component_context can still use it below
+      dir = zx::channel(fdio_service_clone(flat->directories.at(i).get()));
+    } else {
+      dir = std::move(flat->directories.at(i));
+    }
+
+    zx_handle_t dir_handle = dir.release();
+    const char* path = flat->paths.at(i).data();
+    zx_status_t status = fdio_ns_bind(ns, path, dir_handle);
+    if (status != ZX_OK) {
+      FX_LOGF(ERROR, LOG_TAG, "Failed to bind %s to namespace: %s",
+              flat->paths.at(i).c_str(), zx_status_get_string(status));
+      zx_handle_close(dir_handle);
+      return nullptr;
+    }
+  }
+
+  return ns;
+}
+
+int DartComponentControllerV1::GetStdoutFileDescriptor() {
+  return SetupFileDescriptor(std::move(startup_info_.launch_info.out));
+}
+
+int DartComponentControllerV1::GetStderrFileDescriptor() {
+  return SetupFileDescriptor(std::move(startup_info_.launch_info.err));
+}
+
+int DartComponentControllerV1::SetupFileDescriptor(
+    fuchsia::sys::FileDescriptorPtr fd) {
+  if (!fd) {
+    return -1;
+  }
+  // fd->handle1 and fd->handle2 are no longer used.
+  int outfd = -1;
+  zx_status_t status = fdio_fd_create(fd->handle0.release(), &outfd);
+  if (status != ZX_OK) {
+    FX_LOGF(ERROR, LOG_TAG, "Failed to extract output fd: %s",
+            zx_status_get_string(status));
+    return -1;
+  }
+  return outfd;
+}
+
+bool DartComponentControllerV1::PrepareBuiltinLibraries() {
+  auto directory_request =
+      std::move(startup_info_.launch_info.directory_request);
+
+  auto* flat = &startup_info_.flat_namespace;
+  std::unique_ptr<sys::ServiceDirectory> svc;
+  for (size_t i = 0; i < flat->paths.size(); ++i) {
+    zx::channel dir;
+    if (flat->paths.at(i) == kServiceRootPath) {
+      svc = std::make_unique<sys::ServiceDirectory>(
+          std::move(flat->directories.at(i)));
```

Comment:
There is an interesting ordering problem here that I just noticed. If we call PrepareBuiltinLibraries before we call PrepareNamespace we will fail to serve /svc because we use std::move() here whereas in PrepareNamespace we clone this file descriptor. I wonder if it would be safer to clone it in both places or put in some sort of ordering check to make sure this gets called after prepare namespace.